src/transformers/pipelines/base.py

```diff
@@ -739,65 +826,107 @@ def check_model_type(self, supported_models: Union[List[str], dict]):
+        """
+        _forward will receive the prepared dictionnary from `preprocess` and run it on the model. This method might
+        involve the GPU or the CPU and should be agnostic to it. Isolating this function is the reason for `preprocess`
+        and `postprocess` to exist, so that the hot path, this method generally can run as fast as possible.
+
+        It is not meant to be called directly, `forward` is preferred. It is basically the same but contains additional
+        code surrounding `_forward` making sure tensors and models are on the same device, disabling the training part
+        of the code (leading to faster inference).
+        """
+        raise NotImplementedError("_forward not implemented")
+
+    def forward(self, model_inputs, **forward_params):
         with self.device_placement():
             if self.framework == "tf":
-                # TODO trace model
-                predictions = self.model(inputs.data, training=False)[0]
-            else:
+                model_inputs["training"] = False
+                model_outputs = self._forward(model_inputs, **forward_params)
+            elif self.framework == "pt":
                 with torch.no_grad():
-                    inputs = self.ensure_tensor_on_device(**inputs)
-                    predictions = self.model(**inputs)[0].cpu()
-
-        if return_tensors:
-            return predictions
+                    model_inputs = self.ensure_tensor_on_device(**model_inputs)
+                    model_outputs = self._forward(model_inputs, **forward_params)
+                    model_inputs = self.ensure_tensor_on_cpu(model_outputs)
+            else:
+                raise ValueError(f"Framework {self.framework} is not supported")
+        return model_outputs
+
+    def get_iterator(self, inputs, num_workers: int, preprocess_params, forward_params, postprocess_params):
+        if "TOKENIZERS_PARALLELISM" not in os.environ:
+            logger.info("Disabling tokenizer parallelism, we're using DataLoader multithreading already")
+            os.environ["TOKENIZERS_PARALLELISM"] = "false"
+        dataset = PipelineDataset(inputs, self.preprocess, preprocess_params)
+        dataloader = DataLoader(dataset, num_workers=num_workers, batch_size=1, collate_fn=collate_fn)
+        model_iterator = PipelineIterator(dataloader, self.forward, forward_params)
+        final_iterator = PipelineIterator(model_iterator, self.postprocess, postprocess_params)
+        return final_iterator
+
+    def __call__(self, inputs, *args, num_workers=8, **kwargs):
+        if args:
+            logger.warning(f"Ignoring args : {args}")
+        preprocess_params, forward_params, postprocess_params = self._sanitize_parameters(**kwargs)
+
+        # Fuse __init__ params and __call__ params without modifying the __init__ ones.
+        preprocess_params = {**self._preprocess_params, **preprocess_params}
```

Comment:
this means that the __call__ params would everwrite the init params no? Think this is good! Does it maybe make sense to add a warning though when this happens? E.g. something like:

```
if self._preprocess_params.keys() & preprocess_params.keys():
    logger.warning(f"The parameters {self._preprocess_params.keys() & preprocess_params.keys()} have been overwritten by the passed parameters")
```

Reply:
Yes, it will (but it will only `locally` override, contrary to the previous implementation).

```python
pipe = pipeline("text-generation", max_length=20)

pipe("Something") # Using max_length=20
pipe("else", max_length=100) # using max_length=100
pipe("again") # using max_length=20
```

Reply:
I am not a big fan of high verbosity personally but it definitely could be added.
IMO it's expected behavior, so no reason to yell.